nullaway/src/test/java/com/uber/nullaway/NullAwayKeySetIteratorTests.java

```diff
@@ -0,0 +1,132 @@
+
+  @Test
+  public void mapKeySetIteratorBasic() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
+            "public class Test {",
+            "  public void keySetStuff(Map<Object, Object> m) {",
+            "    for (Object k: m.keySet()) {",
+            "      m.get(k).toString();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void mapKeySetIteratorShadowing() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
+            "public class Test {",
+            "  private final Object k = new Object();",
+            "  public void keySetStuff(Map<Object, Object> m) {",
+            "    for (Object k: m.keySet()) {",
+            "      m.get(k).toString();",
+            "    }",
+            "    // BUG: Diagnostic contains: dereferenced expression m.get(k) is @Nullable",
+            "    m.get(k).toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void mapKeySetIteratorDeeperAccessPath() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
+            "public class Test {",
+            "  static class MapWrapper {",
+            "    Map<Object,Object> mf = new HashMap<>();",
+            "    public Map<Object,Object> getMF() { return mf; }",
+            "    public Map<Object,Object> getMF2(int i) { return mf; }",
+            "  }",
+            "  public void keySetStuff(MapWrapper mw, int j) {",
+            "    for (Object k: mw.mf.keySet()) {",
+            "      mw.mf.get(k).toString();",
+            "    }",
+            "    for (Object k: mw.getMF().keySet()) {",
+            "      mw.getMF().get(k).toString();",
+            "    }",
+            "    for (Object k: mw.getMF2(10).keySet()) {",
+            "      mw.getMF2(10).get(k).toString();",
+            "    }",
+            "    for (Object k: mw.getMF2(j).keySet()) {",
+            "      // Report error since we cannot represent mw.getMF2(j).get(k) with an access path",
+            "      // BUG: Diagnostic contains: dereferenced expression mw.getMF2(j).get(k) is @Nullable",
+            "      mw.getMF2(j).get(k).toString();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void unhandledCases() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
+            "public class Test {",
+            "  public void keySetStuff(Map<Object, Object> m) {",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "    m.get(m.keySet().iterator().next()).toString();",
```

Comment:
Is this unhandled or a true positive? Technically the map could be empty....

Reply:
If the Map were empty, this would cause a `NoSuchElementException`, not an NPE.  So I would say our report is a false positive, since it refers to the possibility of an NPE.